client/src/test/java/io/pravega/client/stream/impl/ReaderGroupStateManagerTest.java

```diff
@@ -63,6 +63,56 @@ public MockControllerWithSuccessors(String endpoint, int port, ConnectionFactory
         }
     }
     
+    @Test(timeout = 10000)
+    public void testCompaction() throws ReinitializationRequiredException {
```

Comment:
I think this test case is good as we are primarily trying to address the issue of compaction for reader groups. I'm wondering if we can also unit test the core functionality rather than simply observing whether the high-level behavior is correct. For example, we could, again using this same setup with reader groups, verify that the content of the segment is the one expected and that it includes compacted state. What do you think?

Reply:
If I understand you correctly, your asking if we can have a test on StateSynchronizer that verifies compaction works correctly. If so, we have SynchronizerTest.testCompaction() and StateSynchronizer is implemented using RevisionedStreamClient which has RevisionedStreamClientTest.testMark() and RevisionedStreamClientTest.testSegmentTruncation()

Reply:
Maybe that's enough, help me understand what we have.  In this `testCompaction` case, we are using the state of the reader group to validate the compaction functionality, so we are manipulating the state and interleaving calls to compact the state. The assertions are making sure that the results are as expected.

But, this case is not checking what is being written to the pravega segment. As far as I know, it is possible that this is writing garbage, but the in memory state is still ok. I was suggesting that we make sure that what we write to the Pravega segment as part of compacting is actually correct in addition to making sure that the client can handle compacted state correctly when reading.

Reply:
Ah, ok. I have added verification to the test.